### PR TITLE
Improve the docs about ProxyChecks

### LINKED ISSUE #619: Improve the docs about using ProxyChecks
### Feature Request

.

### PR DESCRIPTION
**Pull request by drknzz**
Closes #619 

## Introduced changes

- Improve the docs about ProxyChecks

image
image

##

- [ ] This PR contains breaking changes

### CODE REVIEW

File: docs/guide.rst [hunk 1]
```diff
@@ -101,16 +101,64 @@ high-level :meth:`Contract.from_address` method to get a contract instance.
...
 
-Check out the code!
+ProxyChecks
+###########
+
+Since the Proxy contracts on StarkNet can have different implementations (especially, since everyone can write their own), there is no single way of checking if some contract is a Proxy contract.
+
+Also, there are two ways of proxying a contract on StarkNet:
+ - make `implementation` in a Proxy contract store a `class hash` of the proxied contract
+ - make `implementation` in a Proxy contract store an `address` of the proxied contract
```
Comment (cptartur) [inline docs/guide.rst:111]: I think this is not 100% correct. Afaik all that is required is to have a `__default__` entrypoint that does some forwarding logic. Where exactly is the implementation stored is not defined I think.

File: docs/guide.rst [hunk 2]
```diff
@@ -101,16 +101,64 @@ high-level :meth:`Contract.from_address` method to get a contract instance.
...
+
+`ProxyChecks` check whether the contract is a Proxy contract.
+They do that by trying to get the `address` or `class_hash` of the implementation.
+
+`Contract.from_address` uses these `ProxyChecks` to fetch the `implementation` (address or class hash) of the proxied contract.
```
Comment (cptartur) [inline docs/guide.rst:116]: Let's swap the order of these sections.

File: docs/guide.rst [hunk 3]
```diff
@@ -101,16 +101,64 @@ high-level :meth:`Contract.from_address` method to get a contract instance.
...
+                self, address: Address, client: Client
+            ) -> Optional[int]:
+                """
+                :return: Implementation class hash of contract being proxied by proxy contract at `address`
+                    given as a parameter or None if implementation does not exist.
+                """
+
+It has two methods:
+ - `implementation_address` - returns the `address` of the proxied contract (implement this if your Proxy contract uses the `address` of another contract as `implementation`)
+ - `implementation_hash` - returns the `class_hash` of the proxied contract (implement this if your Proxy contract uses the `class_hash` of another contract as `implementation`)
```
Comment (cptartur) [inline docs/guide.rst:149]: Maybe let's make it a subsection below the code snippet. And have separate snippets for using built in proxy resolvers and other for custom ones.

Comment (cptartur) [reply docs/guide.rst:149]: Maybe let's also add a `.. note` on that if using custom proxy config, Argent and OZ checks will not be used unless explicitly provided (I think that was the case, correct me if I'm wrong).

File: docs/guide.rst [hunk 4]
```diff
@@ -101,16 +101,64 @@ high-level :meth:`Contract.from_address` method to get a contract instance.
...
 .. codesnippet:: ../starknet_py/tests/e2e/docs/guide/test_resolving_proxies.py
     :language: python
     :dedent: 4
 
 .. note::
 
-    Although :meth:`Contract.from_address()` works like a charm it must perform some calls to get an abi of the contract.
-    If you know the abi statically just use the :ref:`Contract` constructor. It will save some time!
+    :meth:`Contract.from_address()` must perform some calls to get an abi of the contract.
+    If you know the abi statically, just use the :ref:`Contract` constructor.
```
Comment (cptartur) [inline docs/guide.rst:161]: Maybe let's have this on top of the section. And perhaps reword it to something along the lines `If you know the abi of the contract, always prefer creating Contract directly from constructor`, maybe adding the info about performing some calls.

**Review by cptartur — CHANGES_REQUESTED** [review]

File: docs/guide.rst [hunk 5]
```diff
@@ -101,16 +101,64 @@ high-level :meth:`Contract.from_address` method to get a contract instance.
...
+                """
+
+            @abstractmethod
+            async def implementation_hash(
+                self, address: Address, client: Client
+            ) -> Optional[int]:
+                """
+                :return: Implementation class hash of contract being proxied by proxy contract at `address`
+                    given as a parameter or None if implementation does not exist.
+                """
```
[the remark this answers is not in the document]
Comment (drknzz) [reply docs/guide.rst:145]: Hmm, maybe I could just put it into a comment in `starknet_py/tests/e2e/docs/guide/test_resolving_proxies.py` and use `.. codesnippet`. But the syntax coloring might not work then. I will see.

[the remark this answers is not in the document]
Comment (drknzz) [reply docs/guide.rst:145]: Another way is to wrap the acutal ProxyCheck class into `# docs: start`, `# docs: end` and just use the file.

[the remark this answers is not in the document]
Comment (war-in) [reply docs/guide.rst:145]: I like the second idea. @cptartur what do you think?

**Review by drknzz — COMMENTED** [review]
See PR description for current docs.

File: docs/guide.rst [hunk 6]
```diff
@@ -96,21 +96,54 @@ Resolving proxies
...
+Since the Proxy contracts on StarkNet can have different implementations (especially, since everyone can write their own), there is no single way of checking if some contract is a Proxy contract.
+
+There are two main ways of proxying a contract on StarkNet:
+ - forward the calls using ``library_call`` and ``class_hash`` of proxied contract
+ - forward the calls using ``delegate_call`` and ``address`` of proxied contract
+
+:meth:`Contract.from_address <starknet_py.contract.Contract.from_address>` uses ``ProxyChecks`` to fetch the ``implementation`` (address or class hash) of the proxied contract.
+
+**ProxyChecks** check whether the contract is a Proxy contract.
+They do that by trying to get the ``address`` or ``class_hash`` of the implementation.
```
Comment (cptartur) [inline docs/guide.rst:122]:
```suggestion
It does that by trying to get the ``address`` or ``class_hash`` of the implementation.
```

File: docs/guide.rst [hunk 7]
```diff
@@ -96,21 +96,54 @@ Resolving proxies
...
+
+By default, ``proxy_config`` uses a configuration with two **ProxyChecks**:
+ - ArgentProxyCheck - resolves `Argent Proxy <https://github.com/argentlabs/argent-contracts-starknet/blob/b7c4af7462a461386d29551400b985832ba942de/contracts/upgrade/Proxy.cairo>`_.
+ - OpenZeppelinProxyCheck - resolves `OpenZeppelin Proxy <https://github.com/OpenZeppelin/cairo-contracts/blob/d12abf335f5c778fd19d6f99e91c099b40865deb/src/openzeppelin/upgrades/presets/Proxy.cairo>`_.
+
+It's possible to define own ProxyCheck subclass and later pass it to :meth:`Contract.from_address <starknet_py.contract.Contract.from_address>`, so it knows how to resolve the Proxy.
```
Comment (cptartur) [inline docs/guide.rst:128]:
```suggestion
It's possible to define own ProxyCheck implementation and later pass it to :meth:`Contract.from_address <starknet_py.contract.Contract.from_address>`, so it knows how to resolve the Proxy.
```

File: docs/guide.rst [hunk 8]
```diff
@@ -96,21 +96,54 @@ Resolving proxies
...
+
+Since the Proxy contracts on StarkNet can have different implementations (especially, since everyone can write their own), there is no single way of checking if some contract is a Proxy contract.
+
+There are two main ways of proxying a contract on StarkNet:
+ - forward the calls using ``library_call`` and ``class_hash`` of proxied contract
+ - forward the calls using ``delegate_call`` and ``address`` of proxied contract
+
+:meth:`Contract.from_address <starknet_py.contract.Contract.from_address>` uses ``ProxyChecks`` to fetch the ``implementation`` (address or class hash) of the proxied contract.
+
+**ProxyChecks** check whether the contract is a Proxy contract.
```
Comment (cptartur) [inline docs/guide.rst:121]:
```suggestion
**ProxyCheck** check whether the contract is a Proxy contract.
```

File: docs/guide.rst [hunk 9]
```diff
@@ -96,21 +96,54 @@ Resolving proxies
...
+ProxyChecks
+###########
+
+Since the Proxy contracts on StarkNet can have different implementations (especially, since everyone can write their own), there is no single way of checking if some contract is a Proxy contract.
+
+There are two main ways of proxying a contract on StarkNet:
+ - forward the calls using ``library_call`` and ``class_hash`` of proxied contract
+ - forward the calls using ``delegate_call`` and ``address`` of proxied contract
+
+:meth:`Contract.from_address <starknet_py.contract.Contract.from_address>` uses ``ProxyChecks`` to fetch the ``implementation`` (address or class hash) of the proxied contract.
```
Comment (cptartur) [inline docs/guide.rst:119]: It should either be lower case `proxy_checks` or `ProxyCheck`.

**Review by cptartur — COMMENTED** [review]
The way I'd structure this section (this is similar to what @war-in wrote) is:

```
Resolving proxies
---

note about abi and text currently above it

snippet using basic proxy resolving 

ProxyChecks
###

details about custom proxies and snippet using custom proxy resolvers
```

File: starknet_py/proxy/proxy_check.py [hunk 10]
```diff
@@ -11,6 +11,8 @@
 from starknet_py.net.models import Address
 
 
+# fmt: off
+# docs-proxy-check: start
 class ProxyCheck(ABC):
```
Comment (war-in) [inline starknet_py/proxy/proxy_check.py:16]: Why is formatting disabled here? :hmmm:

File: docs/_ext/codesnippet.py [hunk 11]
```diff
@@ -79,7 +79,7 @@ def _set_options(self) -> None:
             "start-after", self.default_start_marker
         )
         self.options["end-before"] = self.options.get(
-            "end-after", self.default_end_marker
+            "end-before", self.default_end_marker
```
Comment (war-in) [inline docs/_ext/codesnippet.py:82]: Was it a bug?

Comment (drknzz) [reply docs/_ext/codesnippet.py:82]: YES 👁️ 👄 👁️

File: starknet_py/proxy/proxy_check.py (hunk 10 above)
Comment (drknzz) [reply starknet_py/proxy/proxy_check.py:16]: This class is showcased in documentation and after adding `# docs-proxy-check` comments above and below, the formatter would insert some 2 blank lines between the last line of the class and `# docs-proxy-check: end`

Comment (war-in) [reply starknet_py/proxy/proxy_check.py:16]: I see, `but` I don't think it is the best solution. That section will not be changed too often but still...

File: docs/_ext/codesnippet.py (hunk 11 above)
Comment (drknzz) [reply docs/_ext/codesnippet.py:82]: Yes, I don't know how the fact that `literalinclude` had to be used instead of `codesnippet` didn't make this bug resolved earlier lol

File: starknet_py/proxy/proxy_check.py (hunk 10 above)
Comment (drknzz) [reply starknet_py/proxy/proxy_check.py:16]: Best alternative is to leave it like this then:

image

File: docs/_ext/codesnippet.py (hunk 11 above)
Comment (war-in) [reply docs/_ext/codesnippet.py:82]: I just thought that you didn't implement that feature in the `codesnippet` 😅

File: starknet_py/proxy/proxy_check.py (hunk 10 above)
Comment (war-in) [reply starknet_py/proxy/proxy_check.py:16]: Not perfect, but probably better than disabling `black`. @cptartur what do you think?

Comment (cptartur) [reply starknet_py/proxy/proxy_check.py:16]: Doesn't look that bad. Couldn't we work around this with inline disables with `# fmt: off`  and `#fmt: on` around the affected lines?

File: docs/guide.rst [hunk 12]
```diff
@@ -95,22 +95,64 @@ This is how we can interact with it:
...
+    :language: python
+    :start-after: docs-proxy-check: start
+    :end-before: docs-proxy-check: end
+
+It has two methods:
+ - `implementation_address` - returns the `address` of the proxied contract (implement this if your Proxy contract uses the `address` of another contract as `implementation`)
+ - `implementation_hash` - returns the `class_hash` of the proxied contract (implement this if your Proxy contract uses the `class_hash` of another contract as `implementation`)
+
+
+Here is the code in practice:
```
Comment (cptartur) [inline docs/guide.rst:148]:
```suggestion
Here is the complete example:
```

**Review by cptartur — APPROVED** [review]

**Review by war-in — APPROVED** [review]
